cardano-db/src/Cardano/Db/Schema.hs

```diff
@@ -151,14 +151,15 @@ share
   TxOut
     txId                TxId                noreference
     index               Word64              sqltype=txindex
-    address             Text
+    address             Text Maybe
     addressHasScript    Bool
     paymentCred         ByteString Maybe    sqltype=hash28type
     stakeAddressId      StakeAddressId Maybe noreference
     value               DbLovelace          sqltype=lovelace
     dataHash            ByteString Maybe    sqltype=hash32type
     inlineDatumId       DatumId Maybe       noreference
     referenceScriptId   ScriptId Maybe      noreference
+    addressDetailId     AddressDetailId Maybe noreference
```

Comment:
I think this needs a new Schema.hs file, instead of editting the default tx_out table. It's fine adding a new table, but eventually we don't want to change the existing tables.

Reply:
It's really only changing `address Text Maybe` by adding a `Maybe` This shouldn't change functionality for how current users query the db? 

Though technically we could store the address in both tables as it's not big in size so this would remain just a `Text`.

This solution means that we don't have to keep multiple schemas updated for what isn't really an excessive feature that can be kept behind a config flag.

Just the more we have schemas the more complex the handling of their functionality within the code will become, especially when more than one schema is involved.

I know currently we run raw sql in the code to handle the extra schema and again the more of these we have the harder it will be to see what is happening from within the codebase.

Ideally we could just make these tables dynamic at runtime but from researching https://stackoverflow.com/questions/78871107/dynamic-tables-at-runtime-in-haskell-persistent there doesn't seem to be a way we can do it with Persistent.

This implementation would be an additional table in the base schema + 1 new `Maybe` field for the pointer to table if needed.

If you think that having another schema is still the best way to go then happy to do it, just wanted to note my apprehensions in doing so 👍

Reply:
It's interesting trying to merge the multiple schemas into 1. We could make a pros-cons list of this approach:
**pros:**
- It makes it easier to introduce and **maintain** 1 schema instead of multiple. More and faster feautures means happier users.
- Users get **exposed** to different feautures directly which may motivate them migrate to a better schema solution for them.

**cons**
- We don't take the most benefit of each implementation **performance** wise. By merging the 2 schemas into 1, we force multiple Null or reduntant fields in a table that is hugely populated, so it takes space.
- Many unused null fields can **confuse** users. So can many fields can are annotated as nullable due to a different config. eg the address of a tx_out should never be null normally for a users that only cares about default behaviour.

I'm afraid I don't have a clear opinion right now. I guess an important question is, is the main objective of this feauture to reduce space (in which case cons outweight) or to improve tx_out queries?

Reply:
@kderme Ok I've managed to do something else. We have 2 new schema files and a "base" schema
[BasicAddress](https://github.com/IntersectMBO/cardano-db-sync/blob/1aae0342269828f1bdd2744be94f62db2ecd9fcb/cardano-db/src/Cardano/Db/Schema/Address/BasicAddress.hs)
[ExtendedAddress](https://github.com/IntersectMBO/cardano-db-sync/blob/1aae0342269828f1bdd2744be94f62db2ecd9fcb/cardano-db/src/Cardano/Db/Schema/Address/ExtendedAddress.hs)

The config uses the type `data AddressType = AddressBasic | AddressExtended` and this flag is then used when we migrate the schemas (this part I need help with as I'm not 100% how our migration is working but technically it should work) something in the lines of:
```Haskell
runMigration migrateBaseCardanoDb
case addressType of
  AddressBasic -> runMigration migrateBasicAddressCardanoDb
  AddressExtended -> runMigration migrateExtendedAddressCardanoDb
```
From there we have to handle all the places that uses `DB.TxOut` to make sure we are using the right types for the right table being used. I converted a couple of queries in slightly different styles in [Cardano.Db.Query.TxOut](https://github.com/IntersectMBO/cardano-db-sync/blob/1aae0342269828f1bdd2744be94f62db2ecd9fcb/cardano-db/src/Cardano/Db/Query/TxOut.hs)

The draw backs being that I'm not sure if you could switch back and forth with changing the config value, handling cases whenever using `DB.TxOut` and dependent on style chosen in `Cardano.Db.Query.TxOut` we could have two functions for each db query related to effected tables. When generating migration files it's probably best to always take priority on using base/basic schemas (should probably rename `BasicAddress` -> `BaseAddress`). Then have the affected changes documented in our docs.

Come to think of something similar could be implemented with `Cardano.Db.Migration.Extra.CosnumedTxOut.Schema` but how we do the migration I'd need pointing as to what happen because it's not clear to me.

The benefit would be existing table schema would be exactly the same and we'd have a "Base" schema with the ability to have custom schemas to handle specific scenarios. We'd get the types for both variations so we'd keep the type safety.

Reply:
> The draw backs being that I'm not sure if you could switch back and forth with changing the config value